transaction/memo.go

```diff
@@ -0,0 +1,134 @@
+package transaction
+
+import (
+	"bytes"
+	"encoding/base64"
+	"fmt"
+	"github.com/gagliardetto/solana-go"
+	solanarpc "github.com/gagliardetto/solana-go/rpc"
+)
+
+const BxMemoMarkerMsg = "Powered by bloXroute Trader Api"
+
+var TraderAPIMemoProgram = solana.MustPublicKeyFromBase58("HQ2UUt18uJqKaQFJhgV9zaTdQxUZjNrsKFgoEDquBkcx")
+
+// CreateTraderAPIMemoInstruction generates a transaction instruction that places a memo in the transaction log
+// Having a memo instruction with signals Trader-API usage is required
+func CreateTraderAPIMemoInstruction(msg string) solana.Instruction {
+	if msg == "" {
+		msg = BxMemoMarkerMsg
+	}
+	buf := new(bytes.Buffer)
+	buf.Write([]byte(msg))
+
+	instruction := &solana.GenericInstruction{
+		AccountValues: nil,
+		ProgID:        TraderAPIMemoProgram,
+		DataBytes:     buf.Bytes(),
+	}
+
+	return instruction
+}
+
+func AddMemo(instructions []solana.Instruction, memoContent string, blockhash solana.Hash, owner solana.PublicKey, privateKeys map[solana.PublicKey]solana.PrivateKey) (string, error) {
+	memo := CreateTraderAPIMemoInstruction(memoContent)
+
+	instructions = append(instructions, memo)
+
+	txnBytes, err := buildFullySignedTxn(blockhash, owner, instructions, privateKeys)
+	if err != nil {
+		return "", err
+	}
+	return base64.StdEncoding.EncodeToString(txnBytes), nil
+}
+
+// AddMemoToSerializedTxn adds memo instruction to a serialized transaction, it's primarily used if the user
+// doesn't want to interact with Trader-API directly
+func AddMemoToSerializedTxn(programIDIndex uint16, txBase64, memoContent string,
+	owner solana.PublicKey, privateKeys map[solana.PublicKey]solana.PrivateKey) (string, error) {
+	signedTxBytes, err := solanarpc.DataBytesOrJSONFromBase64(txBase64)
+	if err != nil {
+		return "", err
+	}
+	unsignedTx := solanarpc.TransactionWithMeta{Transaction: signedTxBytes}
+	solanaTx, err := unsignedTx.GetTransaction()
+	if err != nil {
+		return "", err
+	}
+	progID, err := solanaTx.Message.ResolveProgramIDIndex(programIDIndex)
+	if err != nil {
+		return "", err
+	}
+
+	memo := CreateTraderAPIMemoInstruction(memoContent)
+	var instructions []solana.Instruction
+
+	for _, cmpInst := range solanaTx.Message.Instructions {
+		accounts := cmpInst.ResolveInstructionAccounts(&solanaTx.Message)
+		instructions = append(instructions, &solana.GenericInstruction{
+			AccountValues: accounts,
+			ProgID:        progID,
```

Comment:
this part confuses me –– looks like you should call `ResolveProgramIDIndex` for each instruction